### PR TITLE
Add a page for testers2 to bootcamp

### PR DESCRIPTION
**Pull request by chick**
- Page is 2.6_testers2.ipynb
  - Basic example
  - Queue example, shows fork and join
  - Decoupled GCD using fork and join testing
- Add testers2 dependency to ivy load script
- This makes chisel-bootcamp master point to SNAPSHOTS

### CODE REVIEW

**Comment by chick** [comment]
This is just a first stab, suggestions welcome.

**Comment by ducky64** [comment]
Overall looks good.

Some suggestions:
- in the motivation, some of the statements are aspirational (as in, currently inaccurate):
  - we can't target multiple backends yet, and there's even less work on synthesizing static testvectors
  - it's not yet included in base chisel3
- in the setup, can you simplify the import statements as much as possible?
  - get rid of the `iotesters` imports, since I don't think they're used in executable cells
  - get rid of experimental and use normal `Modules`, unless we want to start pushing MultiIOModules in the rest of the bootcamp
- we might want to delay merging until after Bundle literals makes it in, because it will much improve the syntax
- should document why the `chisel3.tester.RawTester.test` import is needed
- in the first example, perhaps also show sequential logic operations (eg, `clk.step(...)`), to make it more explicit that we also support combinational operations
- Queue `entries` should be the size of the queue (how many elements it can store before exerting backpressure), I don't think it's a delay. Are you thinking of `ShiftRegister`?
- I'd consider doing a simpler example with single-element `enqueue` rather than jumping straight into working with sequences, unless most people only work with sequences (transactional-level testing). I think the sequences make more sense with fork and join.
- I'd reproduce the type signatures of enqueueSeq and dequeueSet in the text, and maybe a short description, just so readers don't need to infer from code
- Consider concurrency as the preferred terminology for `fork` / `join`. Parallelism implies a speedup (which we don't get), whereas concurrency is more of a programming model
- I'd lampshade thread ordering (so readers know that we've thought about it), eg "threads are run in a deterministic order, largely according to their order as specified in code, and certain bug-prone operations that depend on other threads are forbidden with runtime checks". I'd prefer details not to be in the main text at this point, though an expanding box would be fine.
- I'd probably write the GCD test differently to make causality more apparent in the testdriver structure. I'd probably loop through the test values, and for each value: do the enqueue, then fork off a dequeue (assuming you wanted full pipelining - if you didn't you can do a blocking dequeue), and join on all the dequeues. I'd also do the Bundle literal constructors inline with enqueue / dequeue.
- You can chain `initSource().setSourceClock(...)`. And whenever implicit clock resolution happens (as in, pushed to FIRRTL master) we won't even need `setSourceClock` in many cases.

**Comment by ducky64** [comment]
Also, if we get rocket in the bootcamp, it might be worth building some "verification IP" for the various rocket interfaces like TileLink / AXI / etc, in the style of Decoupled.enqueue/dequeue. It's a real-life, moderate-complexity example that probably shows the power better than Decoupled operations.

**Comment by chick** [comment]
- in the motivation, some of the statements are aspirational (as in, currently inaccurate):
  - we can't target multiple backends yet, and there's even less work on synthesizing static testvectors
  - it's not yet included in base chisel3
> Changed to have a current and planned section here.
- in the setup, can you simplify the import statements as much as possible?
  - get rid of the `iotesters` imports, since I don't think they're used in executable cells
  - get rid of experimental and use normal `Modules`, unless we want to start pushing MultiIOModules in the rest of the bootcamp
> iotesters eliminated.

> I would prefer to leave MultiIOModules for now, I think it makes the code cleaner. I regard this page as much as a testers2 tease in addition to its training value
- we might want to delay merging until after Bundle literals makes it in, because it will much improve the syntax
> Bundle literals are built using latest current techniques
- should document why the `chisel3.tester.RawTester.test` import is needed
> Done
- in the first example, perhaps also show sequential logic operations (eg, `clk.step(...)`), to make it more explicit that we also support combinational operations
> Added an additional example using step()
- Queue `entries` should be the size of the queue (how many elements it can store before exerting backpressure), I don't think it's a delay. Are you thinking of `ShiftRegister`?
> Removed this text
- I'd consider doing a simpler example with single-element `enqueue` rather than jumping straight into working with sequences, unless most people only work with sequences (transactional-level testing). I think the sequences make more sense with fork and join.
> Now starts with an example using enqueueNow and expectDequeueNow
- I'd reproduce the type signatures of enqueueSeq and dequeueSet in the text, and maybe a short description, just so readers don't need to infer from code
> Added a table with new methods described
- Consider concurrency as the preferred terminology for `fork` / `join`. Parallelism implies a speedup (which we don't get), whereas concurrency is more of a programming model
> Parallel => Concurrent
- I'd lampshade thread ordering (so readers know that we've thought about it), eg "threads are run in a deterministic order, largely according to their order as specified in code, and certain bug-prone operations that depend on other threads are forbidden with runtime checks". I'd prefer details not to be in the main text at this point, though an expanding box would be fine.
> Added a note to this effect
- I'd probably write the GCD test differently to make causality more apparent in the testdriver structure. I'd probably loop through the test values, and for each value: do the enqueue, then fork off a dequeue (assuming you wanted full pipelining - if you didn't you can do a blocking dequeue), and join on all the dequeues. I'd also do the Bundle literal constructors inline with enqueue / dequeue.
> I'm not sure what you mean here. I think the code is pretty clear. 
- You can chain `initSource().setSourceClock(...)`. And whenever implicit clock resolution happens (as in, pushed to FIRRTL master) we won't even need `setSourceClock` in many cases.
> I have done this. It definitely minimizes the boilerplate. Are there times you would not do this, and if so what do those use case and code look like?

File: 2.6_testers2.ipynb [hunk 1]
```diff
@@ -64,9 +75,9 @@
     "\n",
     "|        | iotesters             | testers2            |\n",
     "| :----  | :---                  | :---                |\n",
-    "| poke   | poke(c.io.in1, 6)     | c.io.in1.poke(6)    |\n",
-    "| peek   | peek(c.io.out1)       | c.io.out1.peek(6)   |\n",
-    "| expect | expect(c.io.out1, 6)  | c.io.out1.expect(6) |\n",
+    "| poke   | poke(c.io.in1, 6)     | c.io.in1.poke(6.W)    |\n",
```
Comment (ducky64) [inline 2.6_testers2.ipynb]: Wait, `.W`? as in width?

File: 2.6_testers2.ipynb [hunk 2]
```diff
@@ -192,8 +229,11 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "## EnqueueSeq and DequeueSeq \n",
-    "Note the usage of `c.in.enqueueSeq` and `c.out.expectDequeueSeq` after some initial boiler plate to set up the decoupled interfaces."
+    "## EnqueueNow and expectDequeueNow\n",
+    "*testers2* has some built in methods for dealing with circuits with Decoupled interfaces in the IOs. In this example we will see how to insert and extract values from the `queue`. \n",
```
Comment (ducky64) [inline 2.6_testers2.ipynb]: I don't know if it's discussed elsewhere but it would be nice to rephrase this to not suggest that this is special case logic - it's a library that anyone can write, which I think is one of the main strengths. Perhaps a link to the GitHub page that shows their implementation?

File: 2.6_testers2.ipynb [hunk 3]
```diff
@@ -211,29 +251,41 @@
...
     "}"
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "> As a quick riddle change the value **100** at line 8 to **300** in the above example and observe that this test no longer passes. What is going on?"
+    "## EnqueueSeq and DequeueSeq \n",
+    "Now we are going to use "
```
Comment (ducky64) [inline 2.6_testers2.ipynb]: Huh?

File: 2.6_testers2.ipynb [hunk 4]
```diff
@@ -247,10 +299,19 @@
...
-    "In this section we will look at running sections of a unit test in parallel.\n",
-    "Look at the use of `fork` in the example below.\n",
-    "The chain `fork` methods allows these blocks of code to be executed at the same time, independent of each other.\n",
-    "The `join` method tells the test to wait until both `fork` blocks have completed before moving on to whatever comes next.\n"
+    "In this section we will look at running sections of a unit test concurrently. In order to do this we will introduce 4 new features of testers2.\n",
+    "\n",
+    "| method | description |\n",
+    "| :---   | :---        |\n",
+    "| fork   | launches a concurrent code block, additional forks can be run concurrently to this one via the .fork appended to end of the code block of the preceeding fork |\n",
+    "| join | re-unites multiple related forks back into a single thread |\n",
```
Comment (ducky64) [inline 2.6_testers2.ipynb]: Perhaps re-phrase more clearly as 'into the calling thread'?

File: 2.6_testers2.ipynb [hunk 5]
```diff
@@ -247,10 +299,19 @@
...
+    "| enqueueSeq | Continues to add (enqueue) elements from the `Seq` to a `Decoupled` input interface, one at a time, until the sequence is exhausted |\n",
```
Comment (ducky64) [inline 2.6_testers2.ipynb]: Conceptually I'd separate these from fork and join, they're also not concurrent operations.

File: 2.6_testers2.ipynb [hunk 6]
```diff
@@ -247,10 +299,19 @@
...
+    "| expectDequeueSeq | Removes (dequeues) elements from a `Decoupled` output interface, one at a time, and compares each one to the next element of the `Seq` |\n",
+    "---\n",
+    "\n",
+    "In the example below two `fork`s are chained together, and then `join`ed. In the first `fork` block the `enqueueSeq` will continue to add elements until exhausted. The second `fork` block will `expectDequeueSeq` on each cycle when data is available.\n",
+    "\n",
+    ">The threads created by fork are run in a deterministic order, largely according to their order as specified in code, and certain bug-prone operations that depend on other threads are forbidden with runtime checks. \n"
```
Comment (ducky64) [inline 2.6_testers2.ipynb]: May be worth noting that only operations which interact through the testers2 simulator API (and in the future, some testers2 communication API) are checked for inter-thread effects. One can use shared variables all they like and testers2 won't know a thing.

File: 2.6_testers2.ipynb [hunk 7]
```diff
@@ -413,10 +442,21 @@
...
    "source": [
-    "Our test looks pretty much the same as the earlier Queue tests. But there's more going on\n",
-    "because the computation take multiple cycles so the input enqueue process is blocked as each\n",
-    "GCD is computed. The good news is that test side of this is simple and consistent across \n",
-    "different Decoupled circuits."
+    "Our test looks pretty much the same as the earlier Queue tests.\n",
+    "But there's more going on because the computation take multiple cycles so the input enqueue process is blocked as each GCD is computed.\n",
+    "The good news is that test side of this is simple and consistent across different Decoupled circuits.\n",
+    "\n",
+    "Also introduced here is the new Chisel3 `Bundle` literal notation. Consider the line\n",
```
Comment (ducky64) [inline 2.6_testers2.ipynb]: We may actually want to introduce this earlier in the bootcamp

File: 2.6_testers2.ipynb (hunk 6 above)
Comment (chick) [reply 2.6_testers2.ipynb]: I think I'll take a crack at explaining this in later work. My goal with this PR is to increase early exposure to testers2

File: 2.6_testers2.ipynb (hunk 7 above)
Comment (chick) [reply 2.6_testers2.ipynb]: Prefer not to take this on at this point

File: 2.6_testers2.ipynb (hunk 5 above)
Comment (chick) [reply 2.6_testers2.ipynb]: Done

File: 2.6_testers2.ipynb (hunk 4 above)
Comment (chick) [reply 2.6_testers2.ipynb]: Done

File: 2.6_testers2.ipynb (hunk 3 above)
Comment (chick) [reply 2.6_testers2.ipynb]: Expanded text here to make sense

File: 2.6_testers2.ipynb (hunk 2 above)
Comment (chick) [reply 2.6_testers2.ipynb]: Done

File: 2.6_testers2.ipynb (hunk 1 above)
Comment (chick) [reply 2.6_testers2.ipynb]: Fixed

File: 2.6_testers2.ipynb (hunk 6 above)
Comment (ducky64) [reply 2.6_testers2.ipynb]: Sure, that's fine.

File: 2.6_testers2.ipynb (hunk 7 above)
Comment (ducky64) [reply 2.6_testers2.ipynb]: Sure, that's fine.
